src/adapter/src/catalog/migrate.rs

```diff
@@ -163,6 +175,314 @@ pub(crate) async fn migrate(
 // Please include the adapter team on any code reviews that add or edit
 // migrations.
 
+/// Copies options from relevant `CREATE SOURCE` statements to any relevant `CREATE SUBSOURCE`
+/// statements. This preps for the eventual removal of these options from the `CREATE SOURCE`
+/// statement.
+fn ast_rewrite_create_subsource_options(
+    tx: &mut Transaction,
+    stmt: &mut Statement<Raw>,
+) -> Result<(), anyhow::Error> {
+    use mz_sql::ast::{
+        CreateSourceConnection, CreateSubsourceOption, CreateSubsourceOptionName,
+        CreateSubsourceStatement, MySqlConfigOptionName, PgConfigOptionName, RawItemName, Value,
+        WithOptionValue,
+    };
+    use mz_storage_types::sources::load_generator::ProtoLoadGeneratorSourceExportStatementDetails;
+    use mz_storage_types::sources::mysql::{
+        ProtoMySqlSourceDetails, ProtoMySqlSourceExportStatementDetails,
+    };
+    use mz_storage_types::sources::postgres::{
+        ProtoPostgresSourceExportStatementDetails, ProtoPostgresSourcePublicationDetails,
+    };
+    use mz_storage_types::sources::proto_source_export_statement_details;
+    use mz_storage_types::sources::ProtoSourceExportStatementDetails;
+    use prost::Message;
+
+    struct Rewriter<'a, 'b> {
+        tx: &'a Transaction<'b>,
+    }
+
+    impl<'ast> VisitMut<'ast, Raw> for Rewriter<'_, '_> {
+        fn visit_create_subsource_statement_mut(
+            &mut self,
+            node: &'ast mut CreateSubsourceStatement<Raw>,
+        ) {
+            match &node.of_source {
+                // Not a source export subsource
+                None => (),
+                Some(source) => {
+                    let details = node
+                        .with_options
+                        .iter()
+                        .find(|o| o.name == CreateSubsourceOptionName::Details);
+                    if details.is_some() {
+                        // this subsource has already had its details written, so it's either new or was already
+                        // migrated
+                        return;
+                    }
+                    info!("migrate: populating subsource details: {:?}", node);
+
+                    let external_reference = node
+                        .with_options
+                        .iter()
+                        .find(|o| o.name == CreateSubsourceOptionName::ExternalReference)
+                        .expect("subsources must have external reference");
+                    let (external_schema, external_table) = match &external_reference.value {
+                        Some(WithOptionValue::UnresolvedItemName(name)) => {
+                            let name_len = name.0.len();
+                            (
+                                name.0[name_len - 2].clone().into_string(),
+                                name.0[name_len - 1].clone().into_string(),
```

Comment:
We need to also retrieve and later check the database part here. Even if it's a dummy value for current sources like `"mysql"` for mysql etc

Reply:
I don't think we can or need to. For mysql subsources the `external_reference` only contains the schema and table name (Sean removed the dummy 'mysql' database thing), and for postgres sources the `external_reference` does include the database name but when matching to the vec of tables in `details` there is no database name on the `ProtoPostgresTableDesc` structs. Since all tables in a given postgres publication must belong to the same database it doesn't matter either.

I will leave a comment to explain this in the code.